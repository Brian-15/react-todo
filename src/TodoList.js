import Todo from './Todo.js';
import NewTodoForm from './NewTodoForm.js'
import { useState } from 'react';
import { v4 as uuid } from 'uuid';

const TodoList = ({ initialData }) => {
    const [ todos, setTodos ] = useState(initialData || []);

    const addTodo = text => {
        setTodos(tasks => [...tasks, { id: uuid(), text }]);
    };

    const removeTask = evt => {
        const targetId = evt.target.parentElement.id;
        setTodos(todos => todos.filter(({ id }) => id !== targetId));
    };

    return (
        <>
            <NewTodoForm addTodo={addTodo} key={uuid()} />
            <ul key={uuid()}>
            {todos.map(({ id, text }) => (
                <Todo id={id} text={text} key={id} removeTask={removeTask} />
            ))}
            </ul>
        </>
    );
};

export default TodoList;