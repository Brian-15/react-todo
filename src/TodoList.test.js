import { render, fireEvent, waitForElementToBeRemoved } from '@testing-library/react';
import TodoList from './TodoList.js';

/** mock uuid with an incrementing id starting from 0
 * from: https://medium.com/@tommybernaciak/a-missing-jest-mock-feature-b89b74d7a09c
 */
 jest.mock('uuid', () => ({
    v4: jest.fn().mockImplementation(
      (function(n) {
        return function() {
          return ++n;
        };
      })(0)
    )
}));

describe('TodoList component', () => {
    test('smoke test: renders component', () => {
        render(<TodoList />);
    });

    test('snapshot test', () => {
        const { asFragment } = render(<TodoList text={'test text'} />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('adds new task', () => {
        const {
            queryAllByTestId,
            queryByTestId,
            queryByText
        } = render(<TodoList />);

        expect(queryAllByTestId('task').length).toBe(0);

        const btn = queryByText('Add Task');
        const input = queryByTestId('text-input');

        fireEvent.change(input, { target: { value: 'TEST' } });
        fireEvent.click(btn);

        expect(queryAllByTestId('task').length).toBe(1);
    });

    test('removes task', () => {
        const {
            queryAllByTestId,
            queryByTestId
        } = render(<TodoList initialData={[{ id: 1, text: 'TEST' }]} />);

        expect(queryAllByTestId('task').length).toBe(1);

        const btn = queryByTestId('remove-btn');
        fireEvent.click(btn);
        
        waitForElementToBeRemoved(queryByTestId('task')).then(() => {
            expect(queryAllByTestId('task').length).toBe(0);
        });
        
    });
});