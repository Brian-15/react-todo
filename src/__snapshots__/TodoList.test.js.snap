// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component snapshot test 1`] = `
<DocumentFragment>
  <form>
    <label
      for="text"
    >
      Text
      <input
        data-testid="text-input"
        name="text"
        value=""
      />
    </label>
    <button>
      Add Task
    </button>
  </form>
  <ul />
</DocumentFragment>
`;
